### ISSUE TITLE
[Feat]: Healthcheck improvements

### ISSUE THREAD

**Issue by EinfachHans** [coollabsio/coolify#2332]
### Description

With reference on #2200, i see currently two great points to improve the current health checks:

1. A notification when the health status of a container changes. A health check & status is great, but it would be great to be informed about the health status changing. This notification should probably not be send for containers that are unhealthy while healthcheck period, because there will be a "deployment failed" notification anyways (?)

2. Adjust the start period to work as in docker healthchecks

> Start period provides initialization time for containers that need time to bootstrap. Probe failure during that period will not be counted towards the maximum number of retries. However, if a health check succeeds during the start period, the container is considered started and all consecutive failures will be counted towards the maximum number of retries.

From https://stackoverflow.com/a/53290016

I'm sadly not a php developer, but i would say this should be quite easy:

[Here](https://github.com/coollabsio/coolify/blob/ede2274816dd6ef5f470f8774fdb6e2a6cfa1d74/app/Jobs/ApplicationDeploymentJob.php#L952) instead of sleeping, the current date should be saved and then [here](https://github.com/coollabsio/coolify/blob/ede2274816dd6ef5f470f8774fdb6e2a6cfa1d74/app/Jobs/ApplicationDeploymentJob.php#L979), if the current time is still within the saved start date + the startPeriod, don't increase the counter and maybe log an information about it.

### Minimal Reproduction (if possible, example repository)

I created this a a bug, to be able to add a bounty on it...

### Version

v4.0.0-beta.291

**Comment by Anshgrover23** [comment]
/attempt #2332

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clwuibofu000elc09sv4v4wwy/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by Anshgrover23** [comment]
@algora-pbc check this pr @EinfachHans @cmer @wutangpaul @AshKyd

**Comment by EinfachHans** [comment]
@Anshgrover23 Awesome you tackled this! For the bounty you have to tackle the second point as well, is this possible? 😊

**Comment by Anshgrover23** [comment]
@EinfachHans yes it is possible I will make new commit by today on this issue for second point also

**Comment by Anshgrover23** [comment]
@EinfachHans done with the changes plss check

**Comment by EinfachHans** [comment]
@Anshgrover23 your commit doesn't look like what i described in this issue 🤔

**Comment by Anshgrover23** [comment]
@EinfachHans i dont understand i have done what is told in the issue that is i have created a notification function when the health status of a container changes and changes docker-compose services timeout period which you have mentioned in point 2

**Comment by Anshgrover23** [comment]
@EinfachHans am i wrong and where plss tell?

**Comment by EinfachHans** [comment]
@Anshgrover23 In your second commit you only add whitespaces. Please see description of this issue about point 2. It's about the current health check implementation and to change it to not await the start period. Everything is described in the issue.

Also i would recommend not to ping anyone to review the pr. Whenever there is time andras will review it i guess 😊

**Comment by Anshgrover23** [comment]
@EinfachHans sorry for making spam tagging everywhere and i have changed two  docker compose.yml file you can see in my latest commit and i have made health check implementation. and kindly can u please check for me all this and then plss reply back if u can ? iam waiting for your review

**Comment by Anshgrover23** [comment]
@EinfachHans i want to understand where i am lagging off

**Comment by EinfachHans** [comment]
Again: Please see the original issue description. I paste it here again, i think it is quite clear:

Adjust the start period to work as in docker healthchecks

> Start period provides initialization time for containers that need time to bootstrap. Probe failure during that period will not be counted towards the maximum number of retries. However, if a health check succeeds during the start period, the container is considered started and all consecutive failures will be counted towards the maximum number of retries.

From https://stackoverflow.com/a/53290016

I'm sadly not a php developer, but i would say this should be quite easy:

[Here](https://github.com/coollabsio/coolify/blob/ede2274816dd6ef5f470f8774fdb6e2a6cfa1d74/app/Jobs/ApplicationDeploymentJob.php#L952) instead of sleeping, the current date should be saved and then [here](https://github.com/coollabsio/coolify/blob/ede2274816dd6ef5f470f8774fdb6e2a6cfa1d74/app/Jobs/ApplicationDeploymentJob.php#L979), if the current time is still within the saved start date + the startPeriod, don't increase the counter and maybe log an information about it.

**Comment by Anshgrover23** [comment]
@EinfachHans ok i understand now i am working on this now sorry for the inconvenience

**Comment by Anshgrover23** [comment]
@EinfachHans check i have made a new commit and complete the instrcutions given in description and sorry for any misunderstanding.

**Comment by Wuemeli** [comment]
Is this still needed?

**Comment by EinfachHans** [comment]
@Wuemeli yes

**Comment by samuel-asleep** [comment]
@EinfachHans @peaklabs-dev can I work on this issue ?

**Comment by agntsprk** [comment]
$/attempt #2332\n\nPlan:\n- Locate current healthcheck loop in `ApplicationDeploymentJob.php` around the referenced lines.\n- Replace the fixed `sleep($startPeriod)` behavior with docker-style semantics: record a `startPeriodStartedAt` timestamp when healthchecks begin; during the start period, ignore failures *unless* a success occurs; once a success occurs (or start period elapses), count failures toward retries.\n- Add clear log lines to show when a failure is ignored due to startPeriod.\n- Add/adjust tests (if available) for the healthcheck retry logic (mock time).\n- Provide a short demo video of a container that fails healthchecks during startPeriod then recovers, showing behavior before/after.

**Comment by agntsprk** [comment]
Submitted draft PR: https://github.com/coollabsio/coolify/pull/8132 (includes /claim #2332 in body). Will add the required short demo video shortly.

**Comment by Kabi10** [comment]
/attempt #2332

Implementation plan:
1. **Health status change notifications** — Add `HealthChanged` notification class fired from `GetContainersStatus` whenever a running container transitions between `:healthy` and `:unhealthy`. Supports all channels (email, Discord, Telegram, Slack, Pushover, webhook). Skips notifications during deployment since deployment already has its own failure path.

2. **Fix start period to match Docker semantics** — Instead of sleeping the full `health_check_start_period` before probing, probing starts immediately. Failures during the start period are logged but do not increment the retry counter. As soon as the container reports `healthy`, deployment succeeds — even if the start period hasn't elapsed. This matches Docker's own `start_period` behavior.

Branch: `fix/healthcheck-improvements` (targeting `next`)

**Comment by Kabi10** [comment]
PR updated with tests: #9590 (previous #9589 was auto-closed due to commit metadata). Reopened with Pest test coverage for the notification class and status transition logic.
